### ISSUE TITLE
My async_receive calls are receiving more bytes than expected, and these bytes don't contain valid (expected) data

### ISSUE THREAD

**Issue by dantamont** [chriskohlhoff/asio#1066]
So, I'll start by saying that I am so frustrated at this point that I am willing to pay for an expert to help me pair program on this, or just get to the bottom of things. I'm sure it is something that I'm doing incorrectly, but I'd like to be wrong about that!

I already have a decent understanding of the Asio framework. I have the beginnings of a networking library implemented, but have run into some infrequent (yet reproducible) failures in the unit tests for my TCP Client/server code. Specifically, the error occurs when I handle `async_receive `operations. Very rarely, I end up receiving far more bytes than I expect to. Think >1400 compared to ~50 bytes. The bytes I receive seem to be nonsense. This causes my program to crash.

I would very much appreciate help from somebody who has a truly deep understanding of the Asio library so that I can get some help tracking down my very strange errors. My best guess is that I have not set up my send/receive buffers correctly, or that I am misunderstanding the way by which `async_receive `populates the buffer. However, the error doesn't occur too often, so I'm worried that it is something trickier than that.

[I've created a single source file that reproduces the error, and it's a bit of a monster, but it does reliably replicate the bug. Please let me know if you're willing and able to help me out on this!](https://pastebin.com/80aQAifi)

**Comment by andrei-datcu** [comment]
Some things to point out:

```
    /// @brief Overrideable logic for the completion handler that is called on asynchronous receive
    virtual void onHandleReceive(Uint64_t bytesTransferred) {
        // Continue receiving
        Uint64_t available = availableBytesForSocketRead();
        if (available >= 60000) {
```
This is incorrect. After the receive operation is completed, you are only guaranteed to get `bytesTransferred` in the buffer you passed, so don't read more than that. Please note that this is a tcp socket, and you are using `async_receive` which may return fewer bytes than actually requested. To quote from the [documentation](https://think-async.com/Asio/asio-1.21.0/doc/asio/reference/basic_stream_socket/async_receive/overload1.html#asio.reference.basic_stream_socket.async_receive.overload1.remarks):
> The receive operation may not receive all of the requested number of bytes. Consider using the [async_read](https://think-async.com/Asio/asio-1.21.0/doc/asio/reference/async_read.html) function if you need to ensure that the requested amount of data is received before the asynchronous operation completes.

Furthermore, this function:
```
    /// @brief Receives multiple messages from buffer
    /// @detail This routine iterates over the buffer based on its capacity, so there is a guarantee
    /// that NumAllowedSimultaneousReceives messages will not be overwritten by incoming messages
    virtual void unpackReceive(Uint64_t bytesTransferred) {
```
will most likely need to use its `bytesTransferred` parameter for the reasons I've mentioned above.

TL;DR

- Use `asio::async_read` if you want to get all the requested bytes at once.
- Regardless of what you end up using **always** check for the `bytesTransferred` parameter
- There are some uses for `.available()`, but it doesn't seem you need it

**Comment by dantamont** [comment]
So, can I buy you a beer or something? It looks like simply replacing with an `asio::async_read` call and requesting the amount of bytes for my header size has solved my issue.
